File: app/src/main/java/com/recursia/popularmovies/di/modules/InteractorModule.java
package com.recursia.popularmovies.di.modules;


import com.recursia.popularmovies.domain.DetailScreenInteractor;
import com.recursia.popularmovies.domain.DetailScreenInteractorImpl;
import com.recursia.popularmovies.domain.FavoriteScreenInteractor;
import com.recursia.popularmovies.domain.FavoriteScreenInteractorImpl;
import com.recursia.popularmovies.domain.MoviesListInteractor;
import com.recursia.popularmovies.domain.MoviesListInteractorImpl;
import com.recursia.popularmovies.domain.MoviesRepository;
import com.recursia.popularmovies.domain.SearchScreenInteractor;
import com.recursia.popularmovies.domain.SearchScreenInteractorImpl;

import dagger.Module;
import dagger.Provides;

@Module(includes = RepositoryModule.class)
public class InteractorModule {

    @Provides
    DetailScreenInteractor provideDetailScreenInteractor(MoviesRepository moviesRepository) {
        return new DetailScreenInteractorImpl(moviesRepository);
    }

    @Provides
    FavoriteScreenInteractor provideFavoriteScreenInteractor(MoviesRepository moviesRepository) {
        return new FavoriteScreenInteractorImpl(moviesRepository);
    }

    @Provides
    MoviesListInteractor provideMoviesListInteractor(MoviesRepository moviesRepository) {
        return new MoviesListInteractorImpl(moviesRepository);
    }

    @Provides
    SearchScreenInteractor provideSearchScreenInteractor(MoviesRepository moviesRepository) {
        return new SearchScreenInteractorImpl(moviesRepository);
    }
}

File: app/src/main/java/com/recursia/popularmovies/di/modules/InteractorModule.kt
package com.recursia.popularmovies.di.modules

import com.recursia.popularmovies.domain.*

import dagger.Module
import dagger.Provides

@Module(includes = [RepositoryModule::class])
class InteractorModule {

    @Provides
    internal fun provideDetailScreenInteractor(
        accountRepository: AccountRepository,
        moviesRepository: MoviesRepository,
        translateRepository: TranslateRepository
    ): DetailScreenInteractor {
        return DetailScreenInteractorImpl(accountRepository, moviesRepository, translateRepository)
    }

    @Provides
    internal fun provideMainScreenInteractor(moviesRepository: MoviesRepository): MainScreenInteractor {
        return MainScreenInteractorImpl(moviesRepository)
    }

    @Provides
    internal fun provideSearchScreenInteractor(moviesRepository: MoviesRepository): SearchScreenInteractor {
        return SearchScreenInteractorImpl(moviesRepository)
    }

    @Provides
    internal fun provideAccountScreenInteractor(accountRepository: AccountRepository): AccountScreenInteractor {
        return AccountScreenInteractorImpl(accountRepository)
    }

    @Provides
    internal fun provideGenresBottomSheetInteractor(moviesRepository: MoviesRepository): GenresBottomSheetInteractor {
        return GenreBottomSheetInteractorImpl(moviesRepository)
    }
}
